java/code/src/com/redhat/rhn/manager/setup/ProductSyncManager.java

```diff
@@ -98,70 +104,134 @@
+        ).collect(Collectors.toList());
+
+        // Retrieve the tree structure
+        final Map<Long, Long> productTreeMap =
+            HibernateFactory.doWithoutAutoFlushing(() -> getProductTreeMap(productsId));
+
+        // Evaluate possible conflicts with existing custom channels
+        final Map<Long, List<String>> conflictsMap =
+            HibernateFactory.doWithoutAutoFlushing(() -> verifyChannelConflicts(productsId));
+
         return productIdents.stream().collect(Collectors.toMap(
-                ident -> ident,
-                ident -> {
-                   try {
-                       addProduct(ident, user);
-                       return Optional.<Throwable>empty();
-                   }
-                   catch (ProductSyncException e) {
-                       return Optional.of(e);
-                   }
-               }
-        ));
+            ident -> ident,
+            ident -> Exceptions.handleByReturning(() -> addProduct(ident, user, productTreeMap, conflictsMap)))
+        );
+    }
+
+    /**
+     * Evaluates if the given product are in conflict with existing custom channels.
+     * @param productsId the scc product ids of the product  to check
+     * @return a map containing the internal product id and the list of channel that are in conflicting
+     */
+    @SuppressWarnings("unchecked")
+    public Map<Long, List<String>> verifyChannelConflicts(List<Long> productsId) {
+        final SelectMode query = ModeFactory.getMode("Product_queries", "verify_channel_conflicts");
+        final DataResult<Map<String, Object>> dataResult = query.execute(productsId);
+        if (CollectionUtils.isEmpty(dataResult)) {
+            return Collections.emptyMap();
+        }
+
+        final Map<Long, List<String>> resultMap = new HashMap<>();
+        for (Map<String, Object> row : dataResult) {
+            final Long productId = (Long) row.get("id");
+
+            List<String> conflictChannels = resultMap.computeIfAbsent(productId, id -> new ArrayList<>());
+            conflictChannels.add((String) row.get("label"));
+        }
+
+        return resultMap;
+    }
+
+    /**
+     * Retrieves the tree structure of the products
+     * @param productsId the scc product ids of the product to extract.
+     *                   Only the base product and its children will be computed.
+     * @return a map specifying for each internal product id which is the internal parent id.
+     */
+    @SuppressWarnings("unchecked")
+    public Map<Long, Long> getProductTreeMap(List<Long> productsId) {
+        final SelectMode query = ModeFactory.getMode("Product_queries", "extract_product_tree");
+        final DataResult<Map<String, Long>> dataResult = query.execute(productsId);
+        if (CollectionUtils.isEmpty(dataResult)) {
+            return Collections.emptyMap();
+        }
+
+        final Map<Long, Long> resultMap = new HashMap<>();
```

Comment:
same as above

```java
dataResult.stream().collect(Collectors.toMap(
                row -> row.get("product_id"), 
                row -> row.get("parent_id")));
```

Reply:
Done :+1:

Reply:
I had to revert this :man_facepalming: `Collectors.toMap()` [does not support null values](https://bugs.openjdk.java.net/browse/JDK-8148463) and since `parent_id` is null for the main product, it will always throw `NullPointerException`. I added a comment to explain that.